### ISSUE TITLE
Make `validate_client_config` enforcing local `root_dir` be optional or configurable

### ISSUE THREAD

**Issue by chipsenkbeil** [neovim/neovim#15323]
### Feature already in Vim?

No

### Pre-read

I've developed a plugin ([distant.nvim](https://github.com/chipsenkbeil/distant.nvim)) that provides support for using neovim to work on remote machines without the need to run neovim on those machines directly. Think of it like the neovim version of [nuclide](https://github.com/facebookarchive/nuclide) or [VS Code Remote Development](https://code.visualstudio.com/docs/remote/remote-overview).

The plugin talks to a binary ([distant](https://github.com/chipsenkbeil/distant)) that acts as a bridge between the local and remote machines. The plugin itself captures BufWriteCmd and BufReadCmd to communicate file changes to/from the remote machine.

### Problem

> TLDR; my `root_dir` is virtual and may only exists on the remote machine and not the local one. This causes validation of the config provided to `vim.lsp.start_client` to fail.

The other support the plugin is trying to offer is talking to LSP instances that are only on the remote machine by means of communicating stdin/stdout between the local and remote machines for an LSP instance. This functionality works, but I've hit a snag in that the current neovim lsp codebase performs a [validation of the `root_dir`](https://github.com/neovim/neovim/blob/3521bf7672d2a33e60e2656043374a1584a0285b/runtime/lua/vim/lsp.lua#L221) to ensure that it is a directory. This is done through `uv.fs_state` and I don't see a way for me to inject an alternative way to validate.

The only workaround I can think of right now is to create those directories temporarily on the local machine just so the check passes, but this isn't guaranteed to work in the case that the local machine blocks creation for some permission reason.

Alternatively, I try to rewrite the entire `lsp.start_client` function, but I'm not sure if all of the needed functionality is publicly available for me to use to do so. That and forking something builtin sounds like a bad time.

### Solution

Ideally, I'd like to either be able to provide an alternative enforcement function or mark validation as 'loose' where it allows through the `root_dir` even if it doesn't show up or isn't a directory on the local machine.

I'm totally open to suggestions on ways to make this work, though. It's very, very close to the experience I want. When I tested on a remote machine that had a matching project directory, LSPs just worked :smile:

**Comment by chipsenkbeil** [comment]
I'm wondering if I could get away with passing something like "/" as the root_dir, but then modify the initialization params with `before_init` to set `rootPath`, `rootUri`, and `workspaceFolders` to the actual directories. Thoughts?

**Comment by chipsenkbeil** [comment]
Looks like I can get away with `before_init`. See https://github.com/chipsenkbeil/distant.nvim/issues/13 for code details.

Going to close this out because a change isn't worth it if we have an already-supported path. :)
